File: backend/app/api/endpoints/creative.py
from fastapi import APIRouter, Depends, HTTPException, Security
from fastapi.security import HTTPAuthorizationCredentials
from sqlalchemy.orm import Session
from pydantic import BaseModel
from typing import Optional, List
import logging

from ...db.session import SessionLocal
from ...models.chapter import Chapter
from ...services.llm import LLMService
from ..deps import security as auth_security, get_current_active_user

logger = logging.getLogger(__name__)

router = APIRouter()

def get_db():
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()

class DraftRequest(BaseModel):
    instructions: str
    context_chapter_id: Optional[str] = None
    context_text: Optional[str] = None

class AnalysisRequest(BaseModel):
    content: str

class StoryboardGridRequest(BaseModel):
    story_text: str

@router.post("/storyboard-grid")
async def generate_storyboard_grid(
    req: StoryboardGridRequest,
    auth: HTTPAuthorizationCredentials = Security(auth_security),
    current_user = Depends(get_current_active_user)
):
    logger.info(f"Received storyboard grid request. Story length: {len(req.story_text)}")
    token = auth.credentials if auth else None
    try:
        result = LLMService.generate_storyboard_grid(req.story_text, token=token)
        return {"prompt": result}
    except Exception as e:
        logger.error(f"Storyboard grid generation failed: {e}")
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/draft")
async def draft_content(
    req: DraftRequest, 
    db: Session = Depends(get_db),
    auth: HTTPAuthorizationCredentials = Security(auth_security),
    current_user = Depends(get_current_active_user) # Ensure auth
):
    logger.info(f"Received draft request. Instructions: {req.instructions[:50]}...")
    context_text = req.context_text if req.context_text else ""
    token = auth.credentials if auth else None
    
    if not context_text and req.context_chapter_id:
        chapter = db.query(Chapter).filter(Chapter.id == req.context_chapter_id).first()
        if chapter:
            # Simple context: Summary of previous chapter + title + current content if available?
            # Actually usually context is what comes BEFORE. 
            context_text = f"Previous Chapter: {chapter.title}\nSummary: {chapter.summary or 'No summary'}"
            logger.info(f"Using context from chapter: {req.context_chapter_id}")
    
    try:
        generated_text = LLMService.generate_draft(req.instructions, context_text, token=token)
        logger.info("Draft generation successful")
        return {"content": generated_text}
    except Exception as e:
        logger.error(f"Draft generation failed: {e}")
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/analyze")
async def analyze_impact(
    req: AnalysisRequest,
    auth: HTTPAuthorizationCredentials = Security(auth_security),
    current_user = Depends(get_current_active_user)
):
    logger.info(f"Received analysis request for content length: {len(req.content)}")
    token = auth.credentials if auth else None
    try:
        result = LLMService.analyze_content(req.content, token=token)
        logger.info("Content analysis successful")
        return result
    except Exception as e:
        logger.error(f"Content analysis failed: {e}")
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/context/{chapter_id}")
async def get_context(chapter_id: str, db: Session = Depends(get_db)):
    chapter = db.query(Chapter).filter(Chapter.id == chapter_id).first()
    if not chapter:
        raise HTTPException(status_code=404, detail="Chapter not found")
    
    # Get all chapters in project to show structure
    project_chapters = db.query(Chapter).filter(Chapter.project_id == chapter.project_id).order_by(Chapter.index).all()
    structure = [{"id": c.id, "title": c.title, "index": c.index} for c in project_chapters]
    
    return {
        "project_structure": structure,
        "focus": chapter.summary or "No summary available",
        "characters": [] # TODO: Implement Character model fetching
    }

File: backend/src/routes/creative.ts
import type { FastifyPluginAsync } from 'fastify';
import { z } from 'zod';
import { db } from '../db/database';
import { LLMService } from '../services/llm';
import { WritingService } from '../services/ai/writing_service';

export const creativeRoutes: FastifyPluginAsync = async (app) => {
  app.post('/storyboard-grid', async (request, reply) => {
    const { story_text } = z
      .object({
        story_text: z.string().trim().min(1),
      })
      .parse(request.body);

    try {
      return { prompt: await LLMService.generateStoryboardGrid(story_text) };
    } catch (error: any) {
      return reply.status(500).send({
        detail: `Storyboard grid generation failed: ${error?.message || String(error)}`,
      });
    }
  });

  app.post('/draft', async (request, reply) => {
    const body = z
      .object({
        instructions: z.string().trim().min(1),
        context_chapter_id: z.string().optional().nullable(),
        context_text: z.string().optional().nullable(),
        project_id: z.coerce.number().int().optional().nullable(),
        chapter_id: z.string().optional().nullable(),
        target_word_count: z.coerce.number().int().positive().optional(),
        apply_metadata: z.boolean().optional().default(true),
        /** If true, append draft to chapter content in DB */
        apply: z.boolean().optional().default(false),
      })
      .parse(request.body);

    try {
      const chapterId = body.chapter_id || body.context_chapter_id || null;
      let projectId = body.project_id ?? null;

      if (!projectId && chapterId) {
        const ch = await db.get(
          'SELECT project_id FROM chapter WHERE id = ?',
          chapterId
        );
        projectId = ch?.project_id ?? null;
      }

      // Enhanced path: layered memory + negative constraints
      if (projectId && chapterId) {
        const result = await WritingService.generateChapterDraft({
          projectId,
          chapterId,
          instructions: body.instructions,
          targetWordCount: body.target_word_count,
          includeExisting: true,
        });

        if (body.apply) {
          const chapter = await db.get(
            'SELECT content FROM chapter WHERE id = ?',
            chapterId
          );
          const merged =
            (chapter?.content ? String(chapter.content) + '\n\n' : '') +
            result.content;
          await db.run(
            'UPDATE chapter SET content = ?, condensed_content = COALESCE(?, condensed_content) WHERE id = ?',
            merged,
            result.condensed ?? null,
            chapterId
          );
        } else if (body.apply_metadata && result.condensed) {
          await WritingService.applyMetadata(chapterId, result.condensed);
        }

        return {
          content: result.content,
          condensed: result.condensed,
          next_plot: result.nextPlot,
        };
      }

      // Legacy simple draft
      let contextText = body.context_text || '';
      if (!contextText && body.context_chapter_id) {
        const chapter = await db.get(
          'SELECT title, summary FROM chapter WHERE id = ?',
          body.context_chapter_id
        );
        if (chapter) {
          contextText = `Previous Chapter: ${chapter.title}\nSummary: ${chapter.summary || 'No summary'}`;
        }
      }

      return {
        content: await LLMService.generateDraft(body.instructions, contextText),
      };
    } catch (error: any) {
      return reply.status(500).send({
        detail: `Draft generation failed: ${error?.message || String(error)}`,
      });
    }
  });

  app.post('/analyze', async (request, reply) => {
    const { content } = z
      .object({
        content: z.string().trim().min(1),
      })
      .parse(request.body);

    try {
      return await LLMService.analyzeContent(content);
    } catch (error: any) {
      return reply.status(500).send({
        detail: `Content analysis failed: ${error?.message || String(error)}`,
      });
    }
  });

  app.post('/consistency', async (request, reply) => {
    const { project_id } = z
      .object({
        project_id: z.coerce.number().int(),
      })
      .parse(request.body);

    try {
      const issues = await WritingService.checkConsistency(project_id);
      return { issues };
    } catch (error: any) {
      return reply.status(500).send({
        detail: `Consistency check failed: ${error?.message || String(error)}`,
      });
    }
  });

  app.post('/impact', async (request, reply) => {
    const body = z
      .object({
        project_id: z.coerce.number().int(),
        chapter_id: z.string().min(1),
        apply: z.boolean().optional().default(true),
      })
      .parse(request.body);

    try {
      const impact = await WritingService.analyzeChapterImpact(
        body.project_id,
        body.chapter_id,
        body.apply !== false
      );
      return impact;
    } catch (error: any) {
      return reply.status(500).send({
        detail: `Impact analysis failed: ${error?.message || String(error)}`,
      });
    }
  });

  app.post('/skill', async (request, reply) => {
    const body = z
      .object({
        project_id: z.coerce.number().int(),
        chapter_id: z.string().min(1),
        skill: z.enum([
          'CINEMATIC_REWRITE',
          'ADD_CONFLICT',
          'REVERSE_PLOT',
        ]),
        technique: z.enum(['montage', 'close_up', 'sensory']).optional(),
        conflictType: z
          .enum(['variable_intrusion', 'extreme_pressure'])
          .optional(),
        intensity: z.enum(['low', 'high']).optional(),
        reversalType: z.enum(['motive_switch', 'character_peel']).optional(),
        targetCharacter: z.string().optional(),
        instructions: z.string().optional().default(''),
        apply: z.boolean().optional().default(false),
      })
      .parse(request.body);

    try {
      let skill: any;
      if (body.skill === 'CINEMATIC_REWRITE') {
        skill = {
          op: 'CINEMATIC_REWRITE',
          technique: body.technique || 'sensory',
          instructions: body.instructions || 'Make it cinematic',
        };
      } else if (body.skill === 'ADD_CONFLICT') {
        skill = {
          op: 'ADD_CONFLICT',
          conflictType: body.conflictType || 'variable_intrusion',
          intensity: body.intensity || 'high',
          instructions: body.instructions,
        };
      } else {
        skill = {
          op: 'REVERSE_PLOT',
          reversalType: body.reversalType || 'motive_switch',
          targetCharacter: body.targetCharacter,
          instructions: body.instructions,
        };
      }

      const content = await WritingService.executeSkill({
        projectId: body.project_id,
        chapterId: body.chapter_id,
        skill,
      });

      if (body.apply) {
        await db.run(
          'UPDATE chapter SET content = ? WHERE id = ?',
          content,
          body.chapter_id
        );
      }

      return { content, applied: Boolean(body.apply) };
    } catch (error: any) {
      return reply.status(500).send({
        detail: `Skill execution failed: ${error?.message || String(error)}`,
      });
    }
  });

  app.get('/context/:chapter_id', async (request, reply) => {
    const { chapter_id } = z
      .object({
        chapter_id: z.string().min(1),
      })
      .parse(request.params);
    const chapter = await db.get('SELECT * FROM chapter WHERE id = ?', chapter_id);
    if (!chapter) {
      return reply.status(404).send({ detail: 'Chapter not found' });
    }

    const [chapters, characters, glossary] = await Promise.all([
      db.all(
        'SELECT id, title, "index", summary, status, condensed_content FROM chapter WHERE project_id = ? ORDER BY "index" ASC',
        chapter.project_id
      ),
      db.all(
        'SELECT id, name, role, description FROM character WHERE project_id = ? ORDER BY id ASC',
        chapter.project_id
      ),
      db.all(
        'SELECT id, term, definition, category FROM glossary WHERE project_id = ? ORDER BY id ASC',
        chapter.project_id
      ),
    ]);

    return {
      project_structure: chapters,
      focus: chapter.summary || 'No summary available',
      characters,
      glossary,
    };
  });
};
